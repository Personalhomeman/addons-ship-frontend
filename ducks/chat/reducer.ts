import { Message, ChatActionTypes } from './types';
import { ChatActions } from './actions';

export interface ChatState {
  messages: Message[];
}

const initialState: ChatState = {
  messages: []
};

export const reducer = (state = initialState, action: ChatActions): ChatState => {
  switch (action.type) {
    case ChatActionTypes.SendMessage:
    case ChatActionTypes.ReceiveMessage:
      return {
        ...state,
        messages: [...state.messages, action.payload]
      };
    default:
      return state;
  }
};
