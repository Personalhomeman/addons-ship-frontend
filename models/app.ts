export type App = {
  appSlug: string;
  plan: string;
  title: string;
  avatarUrl: string;
  projectType: ProjectType;
  androidErrors: string[];
  iosErrors: string[];
  colors?: {
    start?: string;
    end?: string;
  };
};

export type ProjectType =
  | 'xamarin'
  | 'ios'
  | 'osx'
  | 'macos'
  | 'android'
  | 'cordova'
  | 'ionic'
  | 'react-native'
  | 'flutter'
  | 'other';
