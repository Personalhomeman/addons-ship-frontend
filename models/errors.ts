export class RequestError extends Error {
  constructor(public status: number, public statusText: string) {
    super(`${status}: ${statusText}`);
  }
}
