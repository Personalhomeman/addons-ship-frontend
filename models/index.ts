import { NextContext } from 'next';
import { Store } from 'redux';
import { Request } from 'express';

export type AppVersionProps = {
  appSlug: string;
  versionId: string;
};

export const AppVersionPageTabs = ['details', 'devices', 'test', 'activity'];

export type AppVersionPageQuery = AppVersionProps & {
  isPublic?: string;
  selectedTab?: typeof AppVersionPageTabs[number];
};

export type AppPageQuery = {
  appSlug: string;
};

export type AppVersionList = Array<{
  version: string;
  appVersions: AppVersion[];
}>;

export type AppVersion = {
  id: number;
  version: string;
  platform: string;
  buildNumber: number;
  buildSlug: string;
  lastUpdate: Date;
  description: string;
  whatsNew: string;
  minimumOs: string;
  minimumSdk: string;
  packageName: string;
  certificateExpires: Date;
  bundleId: string;
  size: number;
  supportedDeviceTypes: string[];
  distributionType: string;
  iconUrl: string;
  appName: string;
  promotionalText: string;
  keywords: string;
  reviewNotes: string;
  supportUrl: string;
  marketingUrl: string;
  scheme: string;
  configuration: string;
  publicInstallPageURL: string;
};

export interface PageContext extends NextContext {
  store: Store;
  isServer: boolean;
  req: Request;
}

export type Platform = 'ios' | 'android';
