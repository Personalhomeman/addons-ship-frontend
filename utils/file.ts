import fetch from 'node-fetch';

export const uploadFileToS3 = async (file: File | undefined, presignedUrl: string | undefined) => {
  if (!file || !presignedUrl) throw new Error('File or presignedUrl is invalid');

  const options = {
    method: 'put',
    body: file as any
  };

  return fetch(presignedUrl, options).then(resp => {
    if (!resp.ok) throw new Error(`${resp.status}: ${resp.statusText}`);
  });
};
