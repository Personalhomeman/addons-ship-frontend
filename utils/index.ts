import mapKeys from 'lodash/mapKeys';
import mapValues from 'lodash/mapValues';
import camelCase from 'lodash/camelCase';
import snakeCase from 'lodash/snakeCase';
import isPlainObject from 'lodash/isPlainObject';

export const actionTypeCreator = (prefix: string) => (actionType: TemplateStringsArray) => `${prefix}_${actionType}`;

export const camelizeKeys = <T>(input: Object): T =>
  (mapKeys(input, (_, key: string) => camelCase(key)) as unknown) as T;

export const camelizeKeysDeep = <T>(input: Object): T =>
  mapValues(camelizeKeys(input), (value: any) => (isPlainObject(value) ? camelizeKeysDeep(value) : value)) as T;

export const snakifyKeys = <T>(input: Object): T =>
  (mapKeys(input, (_, key: string) => snakeCase(key)) as unknown) as T;

export const snakifyKeysDeep = <T>(input: Object): T =>
  mapValues(snakifyKeys(input), (value: any) => (isPlainObject(value) ? snakifyKeysDeep(value) : value)) as T;
