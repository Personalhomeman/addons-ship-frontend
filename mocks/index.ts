import { AppVersion } from '@/models';
import { TestDevice } from '@/models/test-device';

export const mockAppVersion: AppVersion = {
  id: 123,
  version: '1.0.3',
  platform: 'ios',
  buildNumber: 32,
  buildSlug: '13245ads',
  lastUpdate: new Date('2018-09-22T12:42:31Z'),
  description: 'Some semi-long description',
  whatsNew: 'Some longer whats-new',
  minimumOs: '10.2',
  minimumSdk: 'dunno',
  packageName: 'com.package.name',
  certificateExpires: new Date('2018-09-22T12:42:31Z'),
  bundleId: 'bundle-id',
  size: 4567034,
  supportedDeviceTypes: ['iphone', 'ipad'],
  distributionType: 'development',
  iconUrl: 'http://placekitten.com/160/160',
  appName: 'Standup Timer',
  promotionalText: 'Promotional Text',
  keywords: 'Keywords',
  reviewNotes: 'Review Notes',
  supportUrl: 'https://bitrise.io/support',
  marketingUrl: 'https://bitrise.io/marketing',
  scheme: 'piramid',
  configuration: 'canary',
  publicInstallPageURL: 'https://bitrise.io/app/8b334705d8e78276'
};

export const mockTestDevices: TestDevice[] = [
  { deviceId: 'some-device-id-01', deviceType: 'ios', owner: 'test-user-1' },
  { deviceId: 'some-device-id-02', deviceType: 'android', owner: 'test-user-1' },
  { deviceId: 'some-device-id-03', deviceType: 'ios', owner: 'test-user-2' }
];
