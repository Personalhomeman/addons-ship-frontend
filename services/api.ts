import fetch, { Request, RequestInit, Response } from 'node-fetch';
import { AppVersion } from '@/models';

export type Fetch = (url: string | Request, init?: RequestInit) => Promise<Response>;
export type APIConfig = {
  url: string;
};

class APIService {
  constructor(private fetch: Fetch, private config: APIConfig) {}

  async getAppVersion(appSlug: string, versionId: string): Promise<AppVersion> {
    const resp = await {
      data: {
        id: 123,
        version: '1.0.3',
        platform: 'ios',
        build_number: '32',
        build_slug: '13245ads',
        last_update: '2018-09-22T12:42:31Z',
        description: 'Some semi-long description',
        whats_new: 'Some longer whats-new',
        minimum_os: '10.2',
        minimum_sdk: 'dunno',
        package_name: 'com.package.name',
        certificate_expires: '2018-09-22T12:42:31Z',
        bundle_id: 'bundle-id',
        size: 4567034,
        supported_device_types: ['iphone', 'ipad'],
        distribution_type: 'development',
        icon_url: 'http://placekitten.com/160/160',
        app_name: 'Standup Timer',
        promotional_text: 'Promotional Text',
        keywords: 'Keywords',
        review_notes: 'Review Notes',
        support_url: 'https://bitrise.io/support',
        marketing_url: 'https://bitrise.io/marketing',
        scheme: 'piramid',
        configuration: 'canary'
      }
    };

    const { data } = resp;

    return {
      id: data.id,
      version: data.version,
      platform: data.platform,
      buildNumber: parseInt(data.build_number, 10),
      buildSlug: data.build_slug,
      lastUpdate: new Date(data.last_update),
      description: data.description,
      whatsNew: data.whats_new,
      minimumOs: data.minimum_os,
      minimumSdk: data.minimum_sdk,
      packageName: data.package_name,
      certificateExpires: new Date(data.certificate_expires),
      bundleId: data.bundle_id,
      size: data.size,
      supportedDeviceTypes: data.supported_device_types,
      distributionType: data.distribution_type,
      iconUrl: data.icon_url,
      appName: data.app_name,
      promotionalText: data.promotional_text,
      keywords: data.keywords,
      reviewNotes: data.review_notes,
      supportUrl: data.support_url,
      marketingUrl: data.marketing_url,
      scheme: data.scheme,
      configuration: data.configuration
    };
  }
}

export default new APIService(fetch, {
  url: 'todo: get from config'
});
