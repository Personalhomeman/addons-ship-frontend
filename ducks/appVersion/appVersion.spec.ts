jest.mock('@/services/ship-api');
jest.mock('@/utils/file');

import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store';
import thunk from 'redux-thunk';
import { TestScheduler } from 'rxjs/testing';
import { getType } from 'deox';

import { mockAppVersion, mockUploadedScreenshotResponse } from '@/mocks';
import { AppVersion, AppVersionEvent } from '@/models';
import api, { ShipAPIService } from '@/services/ship-api';

import reducer, {
  fetchAppVersion,
  updateAppVersion,
  uploadScreenshots,
  publishAppVersion,
  pollPublishStatus,
  pollPublishStatusEpic
} from '.';
import { uploadFileToS3 } from '@/utils/file';

describe('appVersion', () => {
  let mockStore: MockStoreCreator, store: MockStoreEnhanced;
  beforeEach(() => {
    mockStore = configureMockStore([thunk.withExtraArgument({ shipApi: api })]);
    store = mockStore();
  });

  describe('reducer', () => {
    it('loads an app version', () => {
      const state = reducer(undefined, fetchAppVersion.complete(mockAppVersion));

      expect(state).toMatchSnapshot();
    });

    test('when publishing started', () => {
      const state = reducer(undefined, publishAppVersion.next() as any);

      expect(state).toMatchSnapshot();
    });

    test('when publishing is done', () => {
      const state = reducer(undefined, publishAppVersion.complete() as any);

      expect(state).toMatchSnapshot();
    });

    test('when publishing had an error', () => {
      const state = reducer(undefined, publishAppVersion.complete() as any);

      expect(state).toMatchSnapshot();
    });

    test('when publish polling has results', () => {
      const state = reducer(
        undefined,
        pollPublishStatus.complete([
          { status: 'in-progress', createdAt: new Date('2019-07-08') },
          { status: 'finished', createdAt: new Date('2019-07-09') }
        ] as AppVersionEvent[])
      );

      expect(state).toMatchSnapshot();
    });
  });

  describe('fetchAppVersion', () => {
    it('fetches an app version', async () => {
      (api.getAppVersion as jest.Mock).mockResolvedValueOnce(mockAppVersion);
      (api.getScreenshots as jest.Mock).mockResolvedValueOnce([mockUploadedScreenshotResponse]);
      await store.dispatch(fetchAppVersion('app-slug', 'version-id') as any);

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't fetch an app version", async () => {
      (api.getAppVersion as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(fetchAppVersion('app-slug', 'version-id') as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('updateAppVersion', () => {
    const updatedAppVersion: AppVersion = { ...mockAppVersion, description: 'Some different description' };

    it('updates an app version', async () => {
      await store.dispatch(updateAppVersion(updatedAppVersion) as any);

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't update an app version", async () => {
      (api.updateAppVersion as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(updateAppVersion(updatedAppVersion) as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('uploadScreenshots', () => {
    it('uploads screenshots', async () => {
      const filename = 'a-file.jpg',
        bits = 'whatever',
        uploadUrl = 'http://some.url?token';
      (api.uploadScreenshots as jest.Mock).mockResolvedValueOnce([
        {
          filename,
          filesize: bits.length,
          uploadUrl
        }
      ]);

      const file = new File([bits], filename);
      await store.dispatch(uploadScreenshots('app-slug', 'version-id', [], [file]) as any);

      expect(api.uploadScreenshots).toHaveBeenCalledWith('app-slug', 'version-id', []);
      expect(uploadFileToS3).toHaveBeenCalledWith(file, uploadUrl);
      expect(api.uploadedScreenshots).toHaveBeenCalledWith('app-slug', 'version-id');
      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't upload screenshots", async () => {
      (api.uploadScreenshots as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(uploadScreenshots('app-slug', 'version-id', [], []) as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('publishAppVersion', () => {
    it('publishes an app version', async () => {
      await store.dispatch(publishAppVersion(mockAppVersion) as any);

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't publishe an app version", async () => {
      (api.publishAppVersion as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(publishAppVersion(mockAppVersion) as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('pollPublishStatus', () => {
    const testScheduler = new TestScheduler((actual, expected) => expect(actual).toEqual(expected));

    it('does not start polling for other actions', () => {
      testScheduler.run(({ cold, expectObservable }) => {
        const action$ = cold('a', { a: { type: 'SOME_OTHER_ACTION' } });

        const output$ = pollPublishStatusEpic(action$ as any, null as any, {} as any);

        expectObservable(output$).toBe('');
      });
    });

    it('starts polling', () => {
      testScheduler.run(({ cold, expectObservable }) => {
        const action$ = cold('a 3s b', {
          a: { type: getType(pollPublishStatus.start), payload: mockAppVersion },
          b: pollPublishStatus.cancel
        });

        const expected = '1s b 999ms b 999ms b';
        const values = {
          a: { type: getType(pollPublishStatus.start) },
          b: { type: getType(pollPublishStatus.complete), payload: ['event1', 'event2'] }
        };

        const dependencies = {
          shipApi: ({
            getAppVersionEvents: () =>
              cold('a', {
                a: ['event1', 'event2']
              })
          } as unknown) as ShipAPIService
        };

        expectObservable(pollPublishStatusEpic(action$, null as any, dependencies)).toBe(expected, values);
      });
    });
  });
});
