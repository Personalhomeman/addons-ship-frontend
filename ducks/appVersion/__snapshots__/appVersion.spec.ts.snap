// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appVersion deleteFeatureGraphic can't delete the feature graphic 1`] = `
Array [
  Object {
    "type": "APP_VERSION_DELETE_FEATURE_GRAPHIC_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_DELETE_FEATURE_GRAPHIC_ERROR",
  },
]
`;

exports[`appVersion deleteFeatureGraphic deletes the feature graphic 1`] = `[Function]`;

exports[`appVersion deleteScreenshot can't delete a screenshot 1`] = `
Array [
  Object {
    "type": "APP_VERSION_DELETE_SCREENSHOTS_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_DELETE_SCREENSHOTS_ERROR",
  },
]
`;

exports[`appVersion deleteScreenshot deletes a screenshot 1`] = `[Function]`;

exports[`appVersion fetchAppVersion can't fetch an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_GET_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_GET_ERROR",
  },
]
`;

exports[`appVersion fetchAppVersion fetches an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_GET_NEXT",
  },
  Object {
    "payload": Object {
      "appName": "Standup Timer",
      "appSlug": "app-slug-123",
      "buildNumber": 32,
      "buildSlug": "13245ads",
      "bundleId": "bundle-id",
      "description": "Some semi-long description",
      "distributionType": "development",
      "iconUrl": "http://placekitten.com/160/160",
      "id": "123",
      "keywords": "Keywords",
      "lastUpdate": 2018-09-22T12:42:31.000Z,
      "marketingUrl": "https://bitrise.io/marketing",
      "minimumOs": "10.2",
      "minimumSdk": "dunno",
      "packageName": "com.package.name",
      "platform": "ios",
      "promotionalText": "Promotional Text",
      "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
      "reviewNotes": "Review Notes",
      "scheme": "piramid",
      "screenshotDatas": Array [
        Object {
          "createdAt": 2018-09-26T12:42:31.000Z,
          "deviceType": "iPhone 6.5‚Äù",
          "downloadUrl": "https://www.bitrise.io/assets/svg/logo-bitrise.svg",
          "filename": "screenshot-1.jpg",
          "filesize": 1000,
          "id": "test-id-1",
          "screenSize": "1024x768",
          "updatedAt": 2018-09-27T12:42:31.000Z,
          "uploaded": true,
        },
      ],
      "supportUrl": "https://bitrise.io/support",
      "supportedDeviceTypes": Array [
        "iphone",
        "ipad",
      ],
      "version": "1.0.3",
      "whatsNew": "Some longer whats-new",
    },
    "type": "APP_VERSION_GET_COMPLETE",
  },
]
`;

exports[`appVersion publishAppVersion can't publishe an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_PUBLISH_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_PUBLISH_ERROR",
  },
]
`;

exports[`appVersion publishAppVersion publishes an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_PUBLISH_NEXT",
  },
  Object {
    "type": "APP_VERSION_PUBLISH_COMPLETE",
  },
]
`;

exports[`appVersion reducer loads an app version 1`] = `
Object {
  "appVersion": Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-123",
    "buildNumber": 32,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": "123",
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "screenshotDatas": Array [],
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.3",
    "whatsNew": "Some longer whats-new",
  },
  "events": Array [],
  "isPublishInProgress": false,
  "isSaving": 0,
}
`;

exports[`appVersion reducer when publish polling has results 1`] = `
Object {
  "appVersion": null,
  "events": Array [
    Object {
      "createdAt": 2019-07-08T00:00:00.000Z,
      "createdAtTimestamp": 1562544000000,
      "status": "in_progress",
      "updatedAt": null,
    },
    Object {
      "createdAt": 2019-07-09T00:00:00.000Z,
      "createdAtTimestamp": 1562630400000,
      "status": "finished",
      "updatedAt": null,
    },
  ],
  "isPublishInProgress": false,
  "isSaving": 0,
}
`;

exports[`appVersion reducer when publishing had an error 1`] = `
Object {
  "appVersion": null,
  "events": Array [],
  "isPublishInProgress": false,
  "isSaving": 0,
}
`;

exports[`appVersion reducer when publishing is done 1`] = `
Object {
  "appVersion": null,
  "events": Array [],
  "isPublishInProgress": false,
  "isSaving": 0,
}
`;

exports[`appVersion reducer when publishing started 1`] = `
Object {
  "appVersion": null,
  "events": Array [],
  "isPublishInProgress": true,
  "isSaving": 0,
}
`;

exports[`appVersion updateAppVersion can't update an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPDATE_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_UPDATE_ERROR",
  },
]
`;

exports[`appVersion updateAppVersion updates an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPDATE_NEXT",
  },
  Object {
    "type": "APP_VERSION_UPDATE_COMPLETE",
  },
]
`;

exports[`appVersion uploadFeatureGraphic can't upload a feature graphic 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPLOAD_FEATURE_GRAPHIC_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_UPLOAD_FEATURE_GRAPHIC_ERROR",
  },
]
`;

exports[`appVersion uploadFeatureGraphic uploads a feature graphic 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPLOAD_FEATURE_GRAPHIC_NEXT",
  },
  Object {
    "type": "APP_VERSION_UPLOAD_FEATURE_GRAPHIC_COMPLETE",
  },
]
`;

exports[`appVersion uploadScreenshots can't upload screenshots 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPLOAD_SCREENSHOTS_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_UPLOAD_SCREENSHOTS_ERROR",
  },
]
`;

exports[`appVersion uploadScreenshots uploads screenshots 1`] = `
Array [
  Object {
    "type": "APP_VERSION_UPLOAD_SCREENSHOTS_NEXT",
  },
  Object {
    "type": "APP_VERSION_UPLOAD_SCREENSHOTS_COMPLETE",
  },
]
`;
